test/distributed/_composable/fsdp/test_fully_shard_training.py

```diff
@@ -1049,6 +1049,82 @@ def _test_2d_mlp_with_nd_mesh(
             self.assertEqual(p.device_mesh.mesh_dim_names, ("dp", "tp"))
 
 
+class TestFullyShardHSDP3DTraining(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(8, torch.cuda.device_count())
+
+    def init_global_mesh(self) -> DeviceMesh:
+        # Prefer to test with >=8 GPUs, but for 2 GPUs, use 2-way TP
+        dp_size = 2 if self.world_size > 2 else 1
+        pp_size = 2 if self.world_size > 4 else 1
+        return init_device_mesh(
+            "cuda",
+            (2, 2, 2),
```

Comment:
nit: I think we need to use the `dp_size` and `pp_size` above.

Reply:
Forgot to remove, but we need 8 GPUs, so it is always 2x2x2. I skip the test if the number of GPUs is less than 8.